pages/agent/v3/hooks.md

```diff
@@ -147,6 +147,14 @@ they are run as part of each job:
 | `pre-exit`      | <span class="add-icon-agent">Agent</span><br /><span class="add-icon-repository">Repository</span><br /><span class="add-icon-plugin">Plugin (non-vendored)</span><br /><span class="add-icon-plugin">Plugin (vendored)</span> | Runs before the job finishes. Useful for performing cleanup tasks. |
 {: class="table table--no-wrap"}
 
+Each `command` job defined in a pipeline's `pipeline.yml` file executes independently, and therefore, a given hook will run for every one of these `command` jobs.
+
+When defining multiple commands in a step using the `commands` attribute, such as the `pipeline.yml` example described in [Command step attributes](/docs/pipelines/command-step#command-step-attributes), then each sub-command is concatenated with `&&` and effectively runs as a single job. Therefore, a given hook will only run once for a given `commands` job, consisting of multiple sub-commands.
+
+> 📘 Hook failure behavior
+> In the table above, if any of the hooks above `command` (from `pre-bootstrap` to `pre-command`, inclusive) fails with a non-zero exit code, then the command job will not run.
+> Since all the hooks below `command` (from `post-command` to `pre-exit`, inclusive) run after the command job, then any non-zero exit code failure in these hooks has no impact on the running of the command job.
```

Comment:
i think the wording here is making things a little bit confusing - we're referring to `command jobs` as the individual commands that get run during the command phase of a given job, but a "command job" is generally what we refer the whole thing - all the phases, all the hooks - as.

the core of this is that hooks after the command hook _will_ affect the command job, because they'll cause the job to fail with whatever status code they exit with. they won't, however, prevent the command itself from running - it's a bit of a nuanced difference, but it's important, i think.

Reply:
Aha! Thanks for explaining this subtlety! This is the distinction I was looking for before! However, prior to this, I did have trouble understanding exactly what constituted a 'command job'.

I'll re-write this now!

Reply:
my figma diagram that i made but decided not to post: the green box is the job, each purple box is a phase, and each reddish postit is a hook
![Uploading Untitled.png…]()

Reply:
my figma diagram that i made but decided not to post: the green box is the job, each purple box is a phase, and each reddish postit is a hook

<img width="1456" alt="Untitled" src="https://github.com/user-attachments/assets/c2f1664c-8f3f-4e28-993b-ddd1fb34ebeb">

Reply:
Thanks @moskyb ! Very helpful.
Hopefully what I've re-written and commited now/since makes sense, but I'm now wondering (from your diagram), if a `post-command` fails, does this also cause the `command` phase to fail too?

Reply:
there's no real concept of a phase failing - they're more conceptual groupings than anything else. the job will fail if any hook returns a nonzero status code, but there's no intermediate "phase failure" that we record or measure.

Reply:
So from a practical perspective @moskyb , is there anything a user can do with a successfully run '`command` phase' but with its subsequent hook execution failing, which would then cause the entire job to fail? I.e. is there anything they can do with they could do with the output of this '`command` phase' execution?

I'm just wanting to know if there's any practical value from a user's perspective in keeping the **Hook failure behavior** note/callout in the docs?

Reply:
If the answer to my question above is 'no', then perhaps we should reword the note/callout to:

> 📘 Hook failure behavior
> If any of the hooks above fails with a non-zero exit code, the entire job will fail.

?

Leaving this for @mbelton-buildkite to finish off and merge in as I will be on leave for the next 3 weeks.

Reply:
Sorry @moskyb , I was in a bit of a rush when I wrote my previous comment above and I missed the significance of what you'd written above - i.e.

> they won't, however, prevent the command itself from running - it's a bit of a nuanced difference, but it's important, i think.

Therefore, what I can see that what I'd written above:

> Any actions conducted by commands and scripts that ran successfully during the job's execution, prior to the point of its failure, will also be aborted.

is now incorrect.

Therefore, I'm committing an update to correctly clarify this callout, and also change it into a heading instead.

Reply:
Thanks heaps for confirming the updates @moskyb !
cc @mbelton-buildkite